### DISCUSSION TITLE
Criteria API で Stream を返す操作について

### DISCUSSION THREAD

**Discussion by ironclad-15256304** [domaframework/doma#1108] [Q&A] [answered]
### 相談内容

いつも大変お世話になっております。

私の参加しているプロジェクトでは、それなりに大きなデータに対する一括処理を行いたい場合があります（ Sprig batch を使用）。
そのようなシチュエーションでは、データの取得を Stream で行い、 Iterator で一件ずつデータを処理するような設計とすることでメモリの消費を抑えたいと考えています。

[Dao でストリームを返す](https://doma.readthedocs.io/ja/latest/query/select/#returning-a-stream) 場合には普通に取得結果が Stream として返されるようでしたので、 Criteria API の NativeSql でも同じだろうと思い、 stream() を使用してデータの取得を行おうと思っていたのですが、ドキュメントを読み込んでいると [フェッチ (Entityql、NativeSql)](https://doma.readthedocs.io/ja/latest/criteria-api/#fetching-entityql-nativesql) に下記のような記載がありました。

> // The stream method returns results as a stream.
> // The following code is equivalent to "entityql.from(e).fetch().stream()"

この記載によると、一旦データを fetch() して、それを stream() で返しているようです。

この違いが生じているのは「 Entityql による重複除外を行うには、一度 fetch しないと正式な結果が返せない」ことからでしょうか？　それとも、単にショートカットとして用意されているだけで深い意味はないのでしょうか。

大量データを処理する場合 SQL の用途が限られるため、動的な処理に強い Criteria API ではなく SQL ファイルで対応しても特に問題はないのですが、 Criteria API で処理可能なものは可能な限り Criteria API で処理したいと考えています。
そのため、リソースをクローズする義務が生じるのは承知の上で、 NativeSql でも取得結果を直接 Stream として返す操作が行えると大変有り難いのですが、そのような操作を実現する方法はありますでしょうか。（ mapStream 等で対応可能なのでしょうか…？）
お忙しいところ大変恐縮ですが、ご回答頂けますと幸いです。

**Comment by nakamura-to** [comment, marked as answer]
[フェッチ (Entityql、NativeSql)](https://doma.readthedocs.io/ja/latest/criteria-api/#fetching-entityql-nativesql) の `stream()` は `fetch().stream()` の単なるショートカットです。
直接Streamを扱うには `mapStream` メソッドを使ってもらう必要があります。
`Stream`を呼び出し元に返すAPIは今のところありません。
`mapStream` メソッドで実現できないか検討してもらえないでしょうか？リソースのクローズはDoma側で実施するのでこちらが安全です。

**Reply by ironclad-15256304** [reply]
承知しました。 `mapStream(stream -> stream)` の形式で実装してみます。
お忙しい中ご回答いただき有難うございます！

**Reply by ironclad-15256304** [reply]
あれ、すみません、ズレたこと書いたような気がします。

> `Stream` を呼び出し元に返すAPIは今のところありません。

と記載がある通り、 `mapStream(stream -> stream)` して `Stream` が返ったところで、その `Stream` は `mapStream` 内でクローズ済みなんですかね？
Spring batch の Chunk を使用してクローズのタイミングは Spring batch 側に任せるような実装の形を想定していますので、自分が実現したいこと自体は `mapStream` では不可能……？なような気がしてきました。

---

もし `mapStream` では実現不可能である場合、将来的に対応して頂いたりすることは可能なのでしょうか？
 Dao と Sql ファイルで対応すればいい話なので強い要望ではないのですが、

1. Web から API で詳細な条件を指定して処理したい（ `List` で取得）
2. Spring batch でざっくりとした範囲を指定して大量に処理したい（ `Stream` で取得）

の２つの要件があるとき、 Criteria API であればわざわざ Sql ファイルを分けずとも同じ文を使いまわして取得方法だけ切り分けられるのではないかと考えていました。

**Reply by nakamura-to** [reply]
mapStreamに関数を渡してもらうことで対応できる可能性もあるのではと思ったのですが無理そうですね。DAOのストリーム検索相当の事ができるように対応したいと思います。

**Reply by ironclad-15256304** [reply]
返信した後で、 "「`mapStream`使えば`Stream`返せます」とは一言も言われてないな…" ということに気づきました。失礼しました。

大変お手数おかけして申し訳ございません。ご検討いただき有難うございます！！

**Reply by nakamura-to** [reply]
#1111 で対応します

**Reply by nakamura-to** [reply]
2.60.0としてリリースしました。
https://github.com/domaframework/doma/releases/tag/2.60.0

**Reply by ironclad-15256304** [reply]
返信が遅くなり申し訳ございません。
迅速なご対応有難うございます。非常に助かります！
